CARDS-1222: Write a script for upgrading the image used by a singular Mongo Docker container

IntegralProgrammer opened the pull request:
This PR introduces the `upgrade_mongo_image.sh` script which:

- Stops a running Docker Compose environment of CARDS which uses a singular (outside of the Docker Compose environment) Docker container of MongoDB for data persistence
- Stops the (outside of Docker Compose environment) MongoDB Docker container
- Pulls the latest `mongo:4.2-bionic` image from Docker Hub
- Restarts the (outside of Docker Compose environment) MongoDB Docker container with the same `/data/db` volume mount as before but with the new `mongo:4.2-bionic` image.
- Restarts the CARDS Docker Compose environment

To test:

- Build this branch (`mvn clean install`)
- `cd compose-cluster`
- `python3 generate_compose_yaml.py --external_mongo --external_mongo_address 172.18.0.1 --external_mongo_dbname sling --subnet 172.18.0.0/16` (leave blank for no database password)
- `docker pull mongo:4.2-bionic`
- `mkdir ~/mongodata`
- `docker run --rm -p 27017:27017 --name mongocards -v $(realpath ~/mongodata):/data/db -d mongo:4.2-bionic`
- `docker-compose build`
- `docker-compose up -d`
- Visit `http://localhost:8080`. Create some data.
- `cd ../Utilities/ContainerManagement`
- `mkdir ~/.docker_tags_backup`
- `./upgrade_mongo_image.sh mongocards`
- Visit `http://localhost:8080`. All previously entered data should still be there.
- Under `~/.docker_tags_backup/` there should be the file `singular_mongo_slingstore.txt` which contains the name and hash of the version of the MongoDB image that was in use before running the `upgrade_mongo_image.sh` script.

IntegralProgrammer committed:
CARDS-1222: Write a script for upgrading the image used by a singular Mongo Docker container

sdumitriu commented on Utilities/ContainerManagement/upgrade_mongo_image.sh:
```diff
@@ -0,0 +1,54 @@
+docker top $MONGO_CONTAINER_INSTANCE 2>/dev/null >/dev/null || { echo "Invalid specified container...exiting."; exit -1; }
+
+# Check that the container has bind-mounted a local directory to /data/db
+DATA_DIRECTORY_MOUNT=$(docker inspect -f '{{json .HostConfig.Binds }}' $MONGO_CONTAINER_INSTANCE | jq -r '.[] | select(test("^.+:/data/db$"))' | cut -d  ':' -f1)
```
Should this also support `...:/data/db:rw` (mount mode at the end)? Right now it doesn't.

IntegralProgrammer replied:
Good idea - I have implemented this in commit `68c1d7c` :heavy_check_mark:

sdumitriu commented on Utilities/ContainerManagement/upgrade_mongo_image.sh:
```diff
@@ -0,0 +1,54 @@
+docker top $MONGO_CONTAINER_INSTANCE 2>/dev/null >/dev/null || { echo "Invalid specified container...exiting."; exit -1; }
+
+# Check that the container has bind-mounted a local directory to /data/db
+DATA_DIRECTORY_MOUNT=$(docker inspect -f '{{json .HostConfig.Binds }}' $MONGO_CONTAINER_INSTANCE | jq -r '.[] | select(test("^.+:/data/db$"))' | cut -d  ':' -f1)
```
Missing `jq` should be handled more gracefully, right now it prints a small error and keeps on processing the script (which, luckily fails, and doesn't mess up the data).

IntegralProgrammer replied:
Implemented in commit `38374ca` :heavy_check_mark:

IntegralProgrammer committed:
CARDS-1222: Write a script for upgrading the image used by a singular Mongo Docker container

Handle when the mount mode of the directory mounted to MONGO_CONTAINER_INSTANCE is specified as RW implicitly (eg. /home/user/mongo:/data/db) as well as when it is specified explicitly (eg. /home/user/mongo:/data/db:rw)

IntegralProgrammer committed:
CARDS-1222: Write a script for upgrading the image used by a singular Mongo Docker container

Exit early if jq is not installed
